### PR TITLE
Enable multiple calls to mbedtls_gcm_update_ad

### LINKED ISSUE #4351: Support chunked AD in gcm.h
With the new `gcm.h` API from https://github.com/ARMmbed/mbedtls/issues/4350, support passing the associated data in chunks, with successive calls to `mbedtls_gcm_update_ad` (directly or via `mbedtls_cipher_update_ad`).

### PR DESCRIPTION
**Pull request by mstarzyk-mobica**
Resolves #4351

### CODE REVIEW

File: tests/suites/test_suite_gcm.function [hunk 1]
```diff
@@ -16,15 +16,19 @@ static int check_multipart( mbedtls_gcm_context *ctx,
     int ok = 0;
     uint8_t *output = NULL;
     size_t n2 = input->len - n1;
+    size_t n1_add = add->len/2;
```
Comment (mstarzyk-mobica) [inline tests/suites/test_suite_gcm.function:19]: since n1 is incremented in a loop and not related to n1_add, I decided to just split the associated data in half.

File: include/mbedtls/gcm.h [hunk 2]
```diff
@@ -69,6 +69,7 @@ typedef struct mbedtls_gcm_context
     int mode;                             /*!< The operation to perform:
                                                #MBEDTLS_GCM_ENCRYPT or
                                                #MBEDTLS_GCM_DECRYPT. */
+    uint8_t ad_data_finished;
```
Comment (mstarzyk-mobica) [inline include/mbedtls/gcm.h:72]: a flag ment to be checked during first call to `mbedtls_gcm_update` or mbedtls_gcm_finish`.

File: library/gcm.c [hunk 3]
```diff
@@ -337,23 +337,36 @@ int mbedtls_gcm_update_ad( mbedtls_gcm_context *ctx,
...
-        return( MBEDTLS_ERR_GCM_BAD_INPUT );
-
-    ctx->add_len = add_len;
+    offset = ctx->add_len % 16;
     p = add;
-    while( add_len > 0 )
+
+    if (offset)
+    {
+        use_len = 16 - offset;
```
Comment (mstarzyk-mobica) [inline library/gcm.c:353]: magic value 16 is used commonly here, probably a good idea to just use the length of `ctx->buf`

File: tests/suites/test_suite_gcm.function (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_gcm.function:19]: Splitting in half is not a good test because the test data lengths don't vary much. The reason `n1` is incremented in a loop is that it gives good test coverage: at the very least, we want the split to happen at 16*a+b for all b=0..15, because there's a risk of bugs related to how far the split is into a 16-byte block.

Correlating `n1` with `n1_add` is not ideal but better than `add->len/2`. Actually I think `add->len - n1` would give slightly better coverage (keeping it within the range `0`..`add->len` of course).

File: include/mbedtls/gcm.h (hunk 2 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/gcm.h:72]: How about using `ctx->len==0`? This tells you whether `update` has been called with a nonzero length, which is slightly different (in case `update` is first called with length=0) but gives the same result in the end. By leveraging this existing information, you don't need the code to update the flag, just the code to finish authenticating the AD when `ctx->len == 0 && ctx->add_len % 16 1= 0` at the beginning of `update` or `finish`.

File: library/gcm.c (hunk 3 above)
Comment (gilles-peskine-arm) [reply library/gcm.c:353]: The length of the buffer is not a good value here: it's the maximum supported block length, not the block length for the algorithm. If we ever support algorithms with a different length (which is not going to happen anytime soon), having `sizeof(buf)` would be worse than `16` since 16 is clearly a magic number but `sizeof(buf)` is correct when initializing and wiping buffers. So please keep 16 as a magic value as in the existing code.

File: include/mbedtls/gcm.h (hunk 2 above)
Comment (mstarzyk-mobica) [reply include/mbedtls/gcm.h:72]: ah! great! exactly what I needed :)

File: library/gcm.c [hunk 4]
```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
         return( MBEDTLS_ERR_GCM_BAD_INPUT );
     }
 
+    if ( ( ctx->len == 0 ) && ( ctx->add_len % 16 ) )
```
Comment (gilles-peskine-arm) [inline library/gcm.c:467]: Style:
```suggestion
    if( ctx->len == 0 && ctx->add_len % 16 != 0 )
```
(No space before `(` per style guide. No implicit comparison with 0 per style guide. Not putting redundant parentheses here is my personal preference, I think they mislead the reader into thinking that something unusual is going on, but feel free to leave them if you prefer.)

File: library/gcm.c [hunk 5]
```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
...
+    {
```
Comment (gilles-peskine-arm) [inline library/gcm.c:468]: Partly on me when I worked on `update` but it's getting worse: please add comment in `gcm.c` (in the documentation of `mbedtls_gcm_context` would be better but it would conflict with #4511 so let's avoid that so let's put it at the top of the implementation of `mbedtls_gcm_update_ad` maybe?) explaining what `buf` contains at various stages:
* When `add_len % 16 == 0` and `len == 0`: initial state.
* When `add_len % 16 != 01 and `len == 0`: the first `add_len % 16` bytes of `buf` have a partial AD block xored in and not yet multiplied in.
* When `len != 0`: the first `add_len % 16` bytes of `buf` have partial ciphertext xored in and not yet multiplied in.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The code looks correct to me. A few documentation changes are needed, and there's an edge case that needs better testing.

File: library/gcm.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply library/gcm.c:479]: - [ ]  add a test function with non-empty AD and empty encrypted data, checking both encryption and decryption, and calling mbedtls_gcm_update(0) 0 or 1 time based on the value of a parameter.

[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply library/gcm.c:479]: Hm. I'm not sure I did what you expected. I added a test in the test data, to cover this scenario. Is it ok?

File: tests/suites/test_suite_gcm.function [hunk 6]
```diff
@@ -16,15 +16,19 @@ static int check_multipart( mbedtls_gcm_context *ctx,
     int ok = 0;
     uint8_t *output = NULL;
     size_t n2 = input->len - n1;
+    size_t n1_add = n1 < add->len ? add->len - n1 : add->len;
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_gcm.function:19]: I am not a big fan of this correlation of the split of additional data and data to encrypt/decrypt. Please add another argument for the split of additionnal data and in the caller functions change to two nested for loops.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/gcm.c [hunk 7]
```diff
@@ -332,28 +332,53 @@ int mbedtls_gcm_starts( mbedtls_gcm_context *ctx,
     return( 0 );
 }
 
-
+/**
+ * mbedtls_gcm_context::buf contains different data type, depending
```
Comment (gilles-peskine-arm) [inline library/gcm.c:336]: It's not “different data type”: it's always the partial state of the computation of the authentication tag. The breakdown is what partially processed input is at different stages.

File: library/gcm.c [hunk 8]
```diff
@@ -332,28 +332,53 @@ int mbedtls_gcm_starts( mbedtls_gcm_context *ctx,
...
+ * on the values of mbedtls_gcm_context::::add_len and
+ * mbedtls_gcm_context::len:
+ *     * When add_len % 16 == 0 and len == 0: initial state.
```
Comment (gilles-peskine-arm) [inline library/gcm.c:339]: Looking at this again I think it would be best explained with four cases:
* If `len == 0 && add_len % 16 == 0`: the first `add_len % 16 == 0` bytes have a partial block of AD that has been xored in but not yet multiplied in.
* If `len == 0 && add_len % 16 == 0`: the authentication tag is correct if the data ends now.
* If `len % 16 != 0`: the first `len % 16 == 0` bytes have a partial block of ciphertext that has been xored in but not yet multiplied in.
* If `len > 0 && len % 16 == 0`: the authentication tag is correct if the data ends now.

File: tests/suites/test_suite_gcm.aes128_de.data [hunk 9]
```diff
@@ -670,6 +670,10 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 depends_on:MBEDTLS_AES_C
 gcm_decrypt_and_verify:MBEDTLS_CIPHER_ID_AES:"659b9e729d12f68b73fdc2f7260ab114":"fd0732a38224c3f16f58de3a7f333da2ecdb6eec92b469544a891966dd4f8fb64a711a793f1ef6a90e49765eacaccdd8cc438c2b57c51902d27a82ee4f24925a864a9513a74e734ddbf77204a99a3c0060fcfbaccae48fe509bc95c3d6e1b1592889c489801265715e6e4355a45357ce467c1caa2f1c3071bd3a9168a7d223e3":"459df18e2dfbd66d6ad04978432a6d97":"ee0b0b52a729c45b899cc924f46eb1908e55aaaeeaa0c4cdaacf57948a7993a6debd7b6cd7aa426dc3b3b6f56522ba3d5700a820b1697b8170bad9ca7caf1050f13d54fb1ddeb111086cb650e1c5f4a14b6a927205a83bf49f357576fd0f884a83b068154352076a6e36a5369436d2c8351f3e6bfec65b4816e3eb3f144ed7f9":32:"8e5a6a79":"FAIL":"":0
 
+AES-GCM NIST Validation (AES-128,128,0,256,128) #0 [#0]
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.aes128_de.data:673]: There are already plenty of test cases with empty ciphertext and non-empty AD, so I don't think the two new test cases are useful.

----

In case they do turn out to be useful:

Please don't add `[#N]` annotations on test cases and avoid adding `#N`. `#N` is a bad historical habit of using numbers rather than useful descriptions. `[#N]` annotations were added automatically when test cases had identical descriptions which made the test reporting hard to use. In both cases, we eventually want to replace those by meaningful descriptions (except in rare cases where we're just testing crypto with some randomish data and there's nothing informative about the difference between randomish data 1 and randomish data 2). If the goal of this test case is to test with empty ciphertext and non-empty AD, the test description should say so. 

Did this test data actually come from a NIST validation suite? I can't find `ff7ad50df57d635839acf143af43f002` in 
http://csrc.nist.gov/groups/STM/cavp/documents/mac/gcmtestvectors.zip .

File: library/gcm.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/gcm.c:479]: I meant to test the scenario where the functions are called in the sequence above. There is currently no test scenario where `mbedtls_gcm_update_ad` is called at least once, but `mbedtls_gcm_update` is not called at all. When we test, even if the plaintext/ciphertext is empty, we call `mbedtls_gcm_update` with a length of 0. It might well make a difference whether `mbedtls_gcm_update` is called at all or not: a plausible bug would be that the last partial block of AD is processed correctly only if `mbedtls_gcm_update` is called at least once, or only if `mbedtls_gcm_update` is called at most once, or only if `mbedtls_gcm_update` is not called at all.

So we need a new test function, which is specialized to empty plaintext/ciphertext, where the AD is a test parameter and where the number of times to call `mbedtls_gcm_update` is a parameter or a loop index. This must be called for at least AD lengths 0, 16 and some number not multiple of 16; for both encryption and decryption; with `mbedtls_gcm_update` called 0, 1 and 2 times.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: library/gcm.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/gcm.c:479]: And similarly we should have tests with empty AD where `mbedtls_gcm_update_ad` is not called at all. Given the current structure of the code, a bug in this case is not likely, but we should test it in case the structure of the code changes.

File: tests/suites/test_suite_gcm.aes128_de.data (hunk 9 above)
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_gcm.aes128_de.data:673]: I was not aware we should pick test vectors from some special place. I guess I'll remove them anyways, since (as you pointed it out) they don't cover increase the coverage.

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:673]: We don't have to pick test vectors from a special place, but 1. we do need to know why we believe that they are correct and 2. if the description says “NIST Validation” then they should indeed come from a NIST validation suite.

**Comment by mstarzyk-mobica** [comment]
1. rebased with removal of redundant commit
```
commit 8fbee92d3cc141d7ca04c7b93e4a854e1d135b1f
Author: Mateusz Starzyk <mateusz.starzyk@mobica.com>
Date:   Wed Jun 9 16:26:55 2021 +0200

    Add gcm tests for empty ciphertext and non-empty AAD.
```

2. rebased to `development` fixing the merge conflict

File: tests/suites/test_suite_gcm.aes128_de.data [hunk 10]
```diff
@@ -670,6 +670,30 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
...
 gcm_decrypt_and_verify:MBEDTLS_CIPHER_ID_AES:"659b9e729d12f68b73fdc2f7260ab114":"fd0732a38224c3f16f58de3a7f333da2ecdb6eec92b469544a891966dd4f8fb64a711a793f1ef6a90e49765eacaccdd8cc438c2b57c51902d27a82ee4f24925a864a9513a74e734ddbf77204a99a3c0060fcfbaccae48fe509bc95c3d6e1b1592889c489801265715e6e4355a45357ce467c1caa2f1c3071bd3a9168a7d223e3":"459df18e2dfbd66d6ad04978432a6d97":"ee0b0b52a729c45b899cc924f46eb1908e55aaaeeaa0c4cdaacf57948a7993a6debd7b6cd7aa426dc3b3b6f56522ba3d5700a820b1697b8170bad9ca7caf1050f13d54fb1ddeb111086cb650e1c5f4a14b6a927205a83bf49f357576fd0f884a83b068154352076a6e36a5369436d2c8351f3e6bfec65b4816e3eb3f144ed7f9":32:"8e5a6a79":"FAIL":"":0
 
+AES-GCM NIST Empty ciphertext. Non-empty AD. Ciphertext updates: 0
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_cipher:MBEDTLS_CIPHER_ID_AES:"63c3f81500746eaf383fe3975d84f849":"0799d4152fd73c1604b4610cf7171fe1":"cb8248e5f904cc9ccccf6f273fe621eee1b4d7ed98480f9e806a48b84e2d6a733772ecf8fb7fe91805715cddab2b462b89f6e6c7cf873f65031f13c357d5f57b00b7c391c39e78ad1ed94be236ca0ae316bce11bc33c5d701fdfc58abbe918b9c42f7b3d6e89d46f9784b388a6e6daf47730b9fa665d755a17e89932fa669c44":"c53d01e53ee4a6ea106ea4a66538265e":0:0
```
Comment (mstarzyk-mobica) [inline tests/suites/test_suite_gcm.aes128_de.data:675]: is it ok to have same test vector for all of these test cases?

File: tests/suites/test_suite_gcm.function [hunk 11]
```diff
@@ -123,11 +124,16 @@ void gcm_encrypt_and_tag( int cipher_id, data_t * key_str,
...
             mbedtls_test_set_step( n1 );
-            if( !check_multipart( &ctx, MBEDTLS_GCM_ENCRYPT,
-                                  iv_str, add_str, src_str,
-                                  dst, tag,
-                                  n1 ) )
-                goto exit;
+
+            for( n1_add = 0; n1_add <= add_str->len; n1_add += 1 )
+            {
+                mbedtls_test_set_step( n1_add );
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.function:201]: The point of the test step is to convey the conditions in which a test failure happens. When tests have no loop or auxiliary functions, the test case name and the line number are enough. The step number provides additional information when the same code line runs more than once for a test case. With a simple loop, the iteration number is enough information. With a nested loop, the step number should encode the step number from both loops. Here, with a test failure, we'd only see the problematic value of `n1_add` and not the problematic value of `n1`.

The step number is printed out in decimal. Following a precedent in `tests/suites/test_suite_ssl.function`:
```
mbedtls_test_set_step( n1 * 10000 + n1_add );
```

Same in the decrypt function.

File: tests/suites/test_suite_gcm.function [hunk 12]
```diff
@@ -123,11 +124,16 @@ void gcm_encrypt_and_tag( int cipher_id, data_t * key_str,
         for( n1 = 0; n1 <= src_str->len; n1 += 1 )
         {
             mbedtls_test_set_step( n1 );
-            if( !check_multipart( &ctx, MBEDTLS_GCM_ENCRYPT,
-                                  iv_str, add_str, src_str,
-                                  dst, tag,
-                                  n1 ) )
-                goto exit;
+
+            for( n1_add = 0; n1_add <= add_str->len; n1_add += 1 )
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.function:199]: The nested loops make the GCM tests a bit slower than I'd like, and we could cut out a lot of values of (n1, n1add) without reducing coverage. But let's optimize later. Filed as https://github.com/ARMmbed/mbedtls/issues/4670.

File: tests/suites/test_suite_gcm.aes128_de.data (hunk 10 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:675]: Yes. The goal is to test the state machine of the GCM calculation. The exact data doesn't matter, only the lengths. For the same reason, it's ok to test only with AES128 and not the other block ciphers.

File: tests/suites/test_suite_gcm.function [hunk 13]
```diff
@@ -196,6 +267,110 @@ exit:
...
+                                          data_t * add_str,
+                                          data_t * tag_str,
+                                          int cipher_update_calls,
+                                          int init_result )
+{
+    mbedtls_gcm_context ctx;
+
+    mbedtls_gcm_init( &ctx );
+
+    TEST_ASSERT( mbedtls_gcm_setkey( &ctx, cipher_id, key_str->x, key_str->len * 8 ) == init_result );
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.function:283]: This function doesn't to exercise the case where setkey fails, so `init_result` is not useful. Also goes for the other 3 functions.

File: tests/suites/test_suite_gcm.aes128_en.data [hunk 14]
```diff
@@ -670,6 +670,30 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 depends_on:MBEDTLS_AES_C
 gcm_encrypt_and_tag:MBEDTLS_CIPHER_ID_AES:"fe481476fce76efcfc78ed144b0756f1":"246e1f2babab8da98b17cc928bd49504d7d87ea2cc174f9ffb7dbafe5969ff824a0bcb52f35441d22f3edcd10fab0ec04c0bde5abd3624ca25cbb4541b5d62a3deb52c00b75d68aaf0504d51f95b8dcbebdd8433f4966c584ac7f8c19407ca927a79fa4ead2688c4a7baafb4c31ef83c05e8848ec2b4f657aab84c109c91c277":"1a2c18c6bf13b3b2785610c71ccd98ca":"b0ab3cb5256575774b8242b89badfbe0dfdfd04f5dd75a8e5f218b28d3f6bc085a013defa5f5b15dfb46132db58ed7a9ddb812d28ee2f962796ad988561a381c02d1cf37dca5fd33e081d61cc7b3ab0b477947524a4ca4cb48c36f48b302c440be6f5777518a60585a8a16cea510dbfc5580b0daac49a2b1242ff55e91a8eae8":"5587620bbb77f70afdf3cdb7ae390edd0473286d86d3f862ad70902d90ff1d315947c959f016257a8fe1f52cc22a54f21de8cb60b74808ac7b22ea7a15945371e18b77c9571aad631aa080c60c1e472019fa85625fc80ed32a51d05e397a8987c8fece197a566689d24d05361b6f3a75616c89db6123bf5902960b21a18bc03a":32:"bd4265a8":0
 
+AES-GCM NIST Empty ciphertext. Non-empty AD. Ciphertext updates: 0
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.aes128_en.data:673]: We know that 16 is a critical value. So please have both a set of test cases where the AD length is a nonzero multiple of 16, and a set where it isn't a multiple of 16. Same with the empty AD tests and the plain/cipher-text length.

File: tests/suites/test_suite_gcm.function [hunk 15]
```diff
@@ -196,6 +267,110 @@ exit:
...
+        check_cipher_with_empty_ad( &ctx, MBEDTLS_GCM_ENCRYPT,
+                                    iv_str, src_str, dst, tag_str,
+                                    ad_update_calls );
+    }
+
+exit:
+    mbedtls_gcm_free( &ctx );
+}
+/* END_CASE */
+
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.function:373]: One more edge case that needs to be tested is when both AD and plaintext/ciphertext are empty, i.e. the sequence is setkey, finish.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
A few remaining issues in the tests. Apart from that looks good to me.

**Review by gilles-peskine-arm — DISMISSED** [review]

File: tests/suites/test_suite_gcm.aes128_de.data [hunk 16]
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
...
+
+AES-GCM NIST - empty ciphertext, AD length: 128 bytes, ciphertext updates: 1
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_cipher:MBEDTLS_CIPHER_ID_AES:"63c3f81500746eaf383fe3975d84f849":"0799d4152fd73c1604b4610cf7171fe1":"cb8248e5f904cc9ccccf6f273fe621eee1b4d7ed98480f9e806a48b84e2d6a733772ecf8fb7fe91805715cddab2b462b89f6e6c7cf873f65031f13c357d5f57b00b7c391c39e78ad1ed94be236ca0ae316bce11bc33c5d701fdfc58abbe918b9c42f7b3d6e89d46f9784b388a6e6daf47730b9fa665d755a17e89932fa669c44":"c53d01e53ee4a6ea106ea4a66538265e":1
+
+AES-GCM NIST - empty ciphertext, AD length: 128 bytes, ciphertext updates: 2
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_cipher:MBEDTLS_CIPHER_ID_AES:"63c3f81500746eaf383fe3975d84f849":"0799d4152fd73c1604b4610cf7171fe1":"cb8248e5f904cc9ccccf6f273fe621eee1b4d7ed98480f9e806a48b84e2d6a733772ecf8fb7fe91805715cddab2b462b89f6e6c7cf873f65031f13c357d5f57b00b7c391c39e78ad1ed94be236ca0ae316bce11bc33c5d701fdfc58abbe918b9c42f7b3d6e89d46f9784b388a6e6daf47730b9fa665d755a17e89932fa669c44":"c53d01e53ee4a6ea106ea4a66538265e":2
+
+AES-GCM NIST - empty ciphertext, AD length: 90 bytes, ciphertext updates: 0
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_gcm.aes128_de.data:685]: Is the reference to NIST correct here ? For the three tests before I can see that there is a corresponding NIST test case before by not for this one. Related question, where does this test data come from? Ditto for the "en" data.

File: tests/suites/test_suite_gcm.aes128_de.data [hunk 17]
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
...
+
+AES-GCM NIST - empty AD, ciphertext length: 128 bytes, AD updates: 2
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_ad:MBEDTLS_CIPHER_ID_AES:"c58583f6479d9bc9f1bffddefee66e59":"cee448b48d3506ff3ecc227a87987846":"564a9f700cbc1f895e4f4fa6426f73b4956896a15e6127e7560d74e3fd0b980d2ee45b7a6a3884fa613d91d13921e3f90967d7132bdafcd146dd8ff7147ed1964c2bdb3e12f4133d3dbbc3bf030ff37b1d2147c493ce885068d9ba5bebae24903aaac004aa0ab73fe789e4150e75ddc2bde2700db02e6398d53e88ac652964ac":"361fc2896d7ee986ecef7cbe665bc60c":"9cce7db3fc087d8cb384f6b1a81f03b3fafa2e3281e9f0fcf08a8283929f32439bb0d302516f0ab65b79181fc223a42345bad6e46ff8bcb55add90207f74481227f71a6230a3e13739ef2d015f5003638234b01e58537b7cfab5a8edac19721f41d46948987d1bb1b1d9485a672647bb3b5cb246a1d753a0d107bff036ac7d95":2
+
+AES-GCM NIST - empty AD, ciphertext length: 51 bytes, AD updates: 0
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_gcm.aes128_de.data:709]: ditto, NIST? Where does the data come from? Ditto for "en" data.

File: tests/suites/test_suite_gcm.aes128_de.data [hunk 18]
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
...
+
+AES-GCM NIST - empty AD, ciphertext length: 51 bytes, AD updates: 2
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_ad:MBEDTLS_CIPHER_ID_AES:"24168b48b45759c8d4f9b061f0cbc16a":"b8e5ede75254cc4542191c7e7b0319ad81651451b639caf81c81c98301a4a0af70e291a4e35b448917be1e400fc64a22edf32913162558c2591ee3e80f397d73dfbc68b82da49bda9bcbb6aaf26919e21c1773cf51f6c5b71784f47978cc0d593b4be0259ab22b0b48de733a884c50a8c148c495973a8f5f84f2e93755666bf5":"be19c7e3d3e63f73d833c967d8d62f388ab9617a2adebe5abd99b5ec64599c46bc28bc62770e08995b0bbf27089e3e17b80424":"4aec633d4daed9ce76d697c11f66f34e":"cb7f10bda7da8a2569ed1f3b667127a1e0fb197283aa16ab8cddd43186bd126b118e671cab3e325877fe0e79f1863f89122c8f":2
+
+AES-GCM NIST - empty AD, empty ciphertext
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_no_ad_no_cipher:MBEDTLS_CIPHER_ID_AES:"cf063a34d4a9a76c2c86787d3f96db71":"113b9785971864c83b01c787":"72ac8493e3a5228b5d130a69d2510e42"
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_gcm.aes128_de.data:723]: Ditto NIST? Origin? "en" data.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
A few remaining things.

File: tests/suites/test_suite_gcm.aes128_de.data (hunk 16 above)
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: Test data comes from http://csrc.nist.gov/groups/STM/cavp/documents/mac/gcmtestvectors.zip

File: tests/suites/test_suite_gcm.function [hunk 19]
```diff
@@ -57,6 +61,94 @@ exit:
...
+    uint8_t *output = NULL;
+    size_t olen;
+
+    /* Sanity checks on the test data */
+    TEST_EQUAL( input->len, expected_output->len );
+
+    TEST_EQUAL( 0, mbedtls_gcm_starts( ctx, mode,
+                                       iv->x, iv->len ) );
+
+    for( n = 0; n < ad_update_count; n += 1 )
```
[the remark this answers is not in the document]
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_gcm.function:82]: `+= 1 ` was actually already used in this file ;)

**Review by gilles-peskine-arm — DISMISSED** [review]

File: tests/suites/test_suite_gcm.aes128_de.data (hunk 16 above)
Comment (ronald-cron-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: Thanks for the reference. Wouldn't it better to say "NIST CAVS 14.0" to identify the source ?

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: It would be nice to identify the source precisely, with the URL, in the commit message. Completely precise identification is a bit too much information in the commit description, which has to fit in 66 characters to be displayed fully.

Comment (ronald-cron-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: Just asking as I don't really know how it works with NIST and its test vectors. Isn't the case that the URL used here could in the future point to a new version of the test vectors which does not include the test vectors added here anymore?

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: Yes. For an unambiguous reference, you need both the URL and some version information. Some indication of the download date, such as an implicit date of “a few days before the commit”, is good enough. A version number is definitely better when there is one.

Comment (mstarzyk-mobica) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: Should I edit the commit messages? (rebase + force push required)

Comment (ronald-cron-arm) [reply tests/suites/test_suite_gcm.aes128_de.data:685]: Editing the commit message is not an issue from our reviewer perspective as we can check that the diff of the force-push is empty. Thus please improve the commit message. Otherwise the NIST files you pointed out start with the header:

 CAVS 14.0
# GCM Decrypt with keysize 128 test information
# Generated on Fri Aug 31 11:28:04 2012

CAVS 14.0 seems to be the version and I think it would be a good idea to mention it in the test description, thus:
`AES-GCM NIST CAVS 14.0 - ... `

**Review by ronald-cron-arm — APPROVED** [review]

File: tests/suites/test_suite_gcm.aes128_en.data [hunk 20]
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 depends_on:MBEDTLS_AES_C
 gcm_encrypt_and_tag:MBEDTLS_CIPHER_ID_AES:"fe481476fce76efcfc78ed144b0756f1":"246e1f2babab8da98b17cc928bd49504d7d87ea2cc174f9ffb7dbafe5969ff824a0bcb52f35441d22f3edcd10fab0ec04c0bde5abd3624ca25cbb4541b5d62a3deb52c00b75d68aaf0504d51f95b8dcbebdd8433f4966c584ac7f8c19407ca927a79fa4ead2688c4a7baafb4c31ef83c05e8848ec2b4f657aab84c109c91c277":"1a2c18c6bf13b3b2785610c71ccd98ca":"b0ab3cb5256575774b8242b89badfbe0dfdfd04f5dd75a8e5f218b28d3f6bc085a013defa5f5b15dfb46132db58ed7a9ddb812d28ee2f962796ad988561a381c02d1cf37dca5fd33e081d61cc7b3ab0b477947524a4ca4cb48c36f48b302c440be6f5777518a60585a8a16cea510dbfc5580b0daac49a2b1242ff55e91a8eae8":"5587620bbb77f70afdf3cdb7ae390edd0473286d86d3f862ad70902d90ff1d315947c959f016257a8fe1f52cc22a54f21de8cb60b74808ac7b22ea7a15945371e18b77c9571aad631aa080c60c1e472019fa85625fc80ed32a51d05e397a8987c8fece197a566689d24d05361b6f3a75616c89db6123bf5902960b21a18bc03a":32:"bd4265a8":0
 
+AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ciphertext updates: 0
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_gcm.aes128_en.data:673]: For future reference, our test framework scripts truncate test case descriptions at 66 characters, so it's better if they fit in 66 characters, and annoying when the first 66 characters aren't unique.
```
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ci  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ci  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ci  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 90 bytes, cip  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 90 bytes, cip  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 90 bytes, cip  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 128 bytes, AD  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 128 bytes, AD  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 128 bytes, AD  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 51 bytes, AD   PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 51 bytes, AD   PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 51 bytes, AD   PASS
```
It's a preexisting problem in many test suites, and we're pressed for time, so I'm not going to block the PR on this.

**Review by gilles-peskine-arm — APPROVED** [review]
